contrib/webpacklib/src/mill/contrib/ScalaJSWebpackModule.scala

```diff
@@ -0,0 +1,166 @@
+          if (byteCount >= 0) {
+            out.write(buffer, 0, byteCount)
+            read(in, buffer, out)
+          } else out
+        }
+
+        val stream: zip.ZipInputStream =
+          new zip.ZipInputStream(
+            new io.BufferedInputStream(new io.FileInputStream(x.path.toIO)))
+        val deps: Seq[JsDeps] =
+          Iterator.continually(stream.getNextEntry)
+            .takeWhile { _ != null }
+            .collect {
+              case z if z.getName == "NPM_DEPENDENCIES" =>
+                JsDeps(ujson.read(read(stream).toString).obj)
+              case z if z.getName.endsWith(".js") &&
+                !z.getName.startsWith("scala/") =>
+                JsDeps(Nil, Nil, Map(z.getName -> read(stream).toString))
+            }
+            .to[Seq]
+        stream.close()
+        deps
+      }
+      .toSeq
+      .foldLeft(JsDeps(Nil, Nil, Map.empty)) { _ ++ _ }
+  }
+
+  def writeBundleSources: Task[(JsDeps, ops.Path) => Unit] = T.task {
+    (jsDeps: JsDeps, dst: ops.Path) =>
+      jsDeps.jsSources foreach { case (n, s) => ops.write(dst / n, s) }
+  }
+
+  def writeWpConfig: Task[(ops.Path, String, String, String, Boolean) => Unit] =
+    T.task {
+      (dst: ops.Path, cfg: String, io: String, out: String, opt: Boolean) =>
+        ops.write(
+          dst / cfg,
+          "module.exports = " + Js.Obj(
+            "mode" -> (if (opt) "production" else "development"),
+            "devtool" -> "source-map",
+            "entry" -> io,
+            "output" -> Js.Obj(
+              "path" -> dst.toString,
+              "filename" -> out)).render(2) + ";\n")
+    }
+
+  def writePackageSpec: Task[(JsDeps, ops.Path) => Unit] = T.task {
+    (jsDeps: JsDeps, dst: ops.Path) =>
+      val compileDeps = jsDeps.compileDependencies
+      val compileDevDeps =
+        jsDeps.compileDevDependencies ++ Seq(
+          "webpack" -> webpackVersion(),
+          "webpack-cli" -> webpackCliVersion(),
+          "webpack-dev-server" -> webpackDevServerVersion(),
+          "source-map-loader" -> "0.2.3")
+
+      ops.write(
+        dst / "package.json",
+        Js.Obj(
+          "dependencies" -> compileDeps,
+          "devDependencies" -> compileDevDeps).render(2) + "\n")
+  }
+
+  def runWebpack: Task[(ops.Path, String) => Unit] = T.task {
+    (dst: ops.Path, cfg: String) =>
+      print(ops.%%("npm", "install")(dst).out.string)
+      print(ops.%%(
+        "node",
+        dst / "node_modules" / "webpack" / "bin" / "webpack",
+        "--bail",
+        "--profile",
+        "--config",
+        cfg)(dst).out.string)
+  }
+
+  def webpack: Task[(ops.Path, ops.Path, Boolean) => Unit] = T.task {
+    (src: ops.Path, dst: ops.Path, opt: Boolean) =>
+      val outjs = dst / src.segments.last
+      val deps = sbtBundlerDeps()
+      val cfg = "webpack.config.js"
+      ops.cp(src, outjs)
+      writeBundleSources().apply(deps, dst)
+      writeWpConfig().apply(dst, cfg, outjs.toString, "out-bundle.js", opt)
```

Comment:
My wishlist:
(1) configurable bundle name and location
(2) ability to provide a path to custom webpack config file

Regarding (2), I think Mill should not be configuring webpack directly. I think letting users configure webpack via plain webpack config files is much more straightforward than via an indirect, limited DSL. Mill's role should be simply to pass any necessary information such as the paths of Scala.js output files to the config.

Reply:
Not having a replacement for scaljs-bundler is just about the only thing keeping me from using mill currently. Having this change merged would be fantastic.

That said, I agree with @raquo that users should at least have the option to configure webpack directly. It's fairly complicated and the plugin ecosystem for webpack is so diverse that a cusom DSL will never keep pace or satisfy all the needs that users will have.

@crimzie , as I said I'm really looking forward to having this committed. Let me know if you could use any help.

Reply:
Hi, @raquo @gavares ! 

I was pretty busy lately and couldn't respond any earlier .

As I wrote in original issue #410, there is a problem of existing libs pulling in dependencies on js sources via sbt-scalajs-bundler. This here is a minimalistic solution for that sole problem. It wasn't originally intended as a full-fledged tool that would let users add their own js dependencies like sbt-scalajs-bundler does and as such doesn't provide customization options.

Yet I agree with your arguments and will rewrite it with an option for user to provide config, just need to find some time to do it :)